src/Symfony/Component/Finder/Iterator/ExcludeDirectoryFilterIterator.php

```diff
@@ -29,10 +31,18 @@ class ExcludeDirectoryFilterIterator extends FilterIterator implements \Recursiv
      */
     public function __construct(\Iterator $iterator, array $directories)
     {
+        $this->iterator = $iterator;
         $this->isRecursive = $iterator instanceof \RecursiveIterator;
-        $this->patterns = array();
+        $patterns = array();
         foreach ($directories as $directory) {
-            $this->patterns[] = '#(^|/)'.preg_quote($directory, '#').'(/|$)#';
+            if (!$this->isRecursive || false !== strpos($directory, '/')) {
+                $patterns[] = preg_quote($directory, '#');
+            } else {
+                $this->excludedDirs[$directory] = true;
+            }
+        }
+        if ($patterns) {
+            $this->excludedPattern = '#(^|/)('.implode('|', $patterns).')(/|$)#';
```

Comment:
I suggest using non-capturing groups here. It will make the regex matching more efficient (the regex engine will not need to capture the groups)